### PR TITLE
Add memory pressure monitoring

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
This is a memory monitoring interface implemented with reference to Linux's PSI (Pressure Stall Information),
which can send notifications when the system's remaining memory is below the threshold.

The following example code sets two different thresholds.
When the system memory is below 10MB, a notification is triggered.
When the system memory is below 20 MB, a notification (POLLPRI event) is triggered every 1s.

```
int main(int argc, FAR char *argv[])
{
  struct pollfd fds[2];
  int ret;

  if (argc == 2)
    {
      char *ptr = malloc(1024*1024*atoi(argv[1]));
      printf("Allocating %d MB\n", atoi(argv[1]));
      ptr[0] = 0;
      return 0;
    }

  fds[0].fd = open("/proc/pressure/memory", O_RDWR);
  fds[1].fd = open("/proc/pressure/memory", O_RDWR);
  fds[0].events = POLLPRI;
  fds[1].events = POLLPRI;

  dprintf(fds[0].fd, "%llu -1", 1024LLU*1024 * 10);
  dprintf(fds[1].fd, "%llu 1000000", 1024LLU*1024 * 20);

  while (1)
    {
      ret = poll(fds, 2, -1);
      if (ret > 0)
        {
          printf("Memory pressure: POLLPRI, %d\n", ret);
        }
    }

  return 0;
}
```

https://docs.kernel.org/accounting/psi.html

### CODE REVIEW

File: arch/sim/src/sim/sim_heap.c [hunk 1]
```diff
@@ -611,6 +611,32 @@ void up_allocate_heap(void **heap_start, size_t *heap_size)
...
+ * Name: mm_heapfree
+ *
+ * Description:
+ *   Return the total free size (in bytes) in the heap
+ *
+ ****************************************************************************/
+
+size_t mm_heapfree(struct mm_heap_s *heap)
+{
+  return SIZE_MAX;
```
Comment (acassis) [inline arch/sim/src/sim/sim_heap.c:624]: Why is it returning a fixed size?

File: arch/sim/src/sim/sim_heap.c [hunk 2]
```diff
@@ -611,6 +611,32 @@ void up_allocate_heap(void **heap_start, size_t *heap_size)
...
+ * Name: mm_heapfree_largest
+ *
+ * Description:
+ *   Return the largest chunk of contiguous memory in the heap
+ *
+ ****************************************************************************/
+
+size_t mm_heapfree_largest(FAR struct mm_heap_s *heap)
+{
+  return SIZE_MAX;
```
Comment (acassis) [inline arch/sim/src/sim/sim_heap.c:637]: Ditto

**Comment by acassis** [comment]
@Gary-Hobson Please include Documentation/ about this new feature

File: arch/sim/src/sim/sim_heap.c (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/sim_heap.c:624]: since this implementation is for sim platform.

File: arch/sim/src/sim/sim_heap.c (hunk 2 above)
Comment (xiaoxiang781216) [reply arch/sim/src/sim/sim_heap.c:637]: ditto

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
> @Gary-Hobson Please include Documentation/ about this new feature

please create a new patch for document

**Pull request merged**